### ISSUE TITLE
Change how stateful material properties are requested

### ISSUE THREAD

**Issue by bwspenc** [idaholab/moose#7867]
### Description of the enhancement or error report

Currently, material properties are made stateful when the model that computes them declares both the new and old property (using `declareProperty()` and `declarePropertyOld()`, for instance). The old (or older) property can be either used in the calculation of that new property or used by some other model, which can request that using `getMaterialPropertyOld()`.

The problem with this system is that the determination of whether an old version of the property is stored is based on whether it was declared, not on whether something is using it. If a model calls `getMaterialPropertyOld()` but that old property wasn't declared, the code happily allows that model to do so, but it gets garbage. We have models that need old state, but the property might be created using a parsed material, which has no way of creating an old state.

This system should be redesigned so that the determination of whether the old state should be stored is based on a request for that old state, not on whether it was declared.
### Rationale for the enhancement or information for reproducing the error

The modular material system in tensor_mechanics relies heavily on transferring data between models using material properties. It has provisions for multiple types of models, including incremental and total models, some of which need old state and some of which don't. A given Material class doesn't necessarily know whether something else will need the old state of a property, and we don't want to store that if it isn't needed (because of issues such as memory usage, restart, and adaptivity).

This system needs to work under the following scenarios:
1. A Material that declares a property needs its old state to compute the current state.
2. A Material that declares a property does not need its old state, but another model does.
3. A model needs to get an old property from another model, which may or may not need the old state of that property itself.
### Identified impact

We discussed this somewhat, and I think this can be accomplished doing something like the following:
- Make the decision on whether to make a material stateful based on whether `getMaterialPropertyOld()` (or Older) is called.
- Get rid of `declarePropertyOld()`. If a Material needs the old state of a property to compute the new state of that property, it would call `getMaterialPropertyOld()` just like a model external to that Material would.

The benefits of this would be:
- Old material properties would only be stored if they are needed.
- Any model requesting an old property would be guaranteed to get a valid value.
- It would not be possible to write to an old property because there would not be a way to get a writeable reference to that property. That would make stateful properties less error-prone.

**Comment by bwspenc** [comment]
Pinging @permcody and @dschwen

**Comment by YaqiWang** [comment]
I just had some thoughts on this and would like to share with you and one more request (not sure if it is doable).

Let us say if a material property is just a function of a variable, although the material is actually stateful, but the developer may rather declare two material properties, one for old another for current and evaluate them both on the fly. I called this pseudo stateful. This is the case where storage is not required but updating old property is required in material. I agree we do not necessarily mess this up with the stateful property.

Another thing is that strictly speaking, current storage is not necessary if a material property is evaluated only once per time step. When there are multiple evaluations, we do not know which one is the last and the old property is no longer needed and can be overwritten. If we can avoid storing the current property, it could save tremendous amount memory.

We have another use case: we do not require the copy for the old state, but require the storage of a material property to avoid possible repeated evaluations. To do this with the current system, we have to have two copies. It will be great that the new design can enable this as well. Possibly the material itself can call `getMaterialPropertyOld()` for the material properties declared by itself, which returns a writeable reference.

So imho, I agree @bwspenc's proposal and want to discuss if we can get rid of the current storage.

**Comment by permcody** [comment]
I'm very much in support of Ben's idea. It actually makes a lot of sense to me and seems like a way superior design over the current system. Consumers really should be controlling whether or not stateful data is needed, not the materials themselves. There are certainly a few gotchas in initialization if properties can be made stateful outside of the knowledge of a material. We'll have to handle this case.

Yaqi, your issues are a little more about optimization or additonal cases that might be encoutered. I'm glad we've captured those cases here so we can consider them as we evolve the current system. Nobody is bringing up adaptivity here but until we have redistribution of properties implemented, many of your issues are probably lower priority. We really need that part to work moving forward.

**Comment by friedmud** [comment]
I like @bwspenc 's idea.  It makes sense.

**Comment by rwcarlsen** [comment]
fixed via #8278
